hledger-lib/Text/Megaparsec/Custom.hs

```diff
@@ -129,6 +151,136 @@ customParseErrorPretty source err = case findCustomError err of
+  , errorSource     :: Text
+  , sourceFileStack :: NE.NonEmpty FilePath
+  } deriving (Show)
+
+type FinalParseErrorBundle = FinalParseErrorBundle' CustomErr
+
+--- * Constructing and throwing final parse errors
+
+-- | Convert a "regular" parse error into a "final" parse error.
+
+errorFinal :: ParseError Char e -> FinalParseError' e
+errorFinal = ErrorFinal
+
+-- | Like 'fancyFailure', but as a "final" parse error.
+
+finalFancyFailure
+  :: (MonadParsec e s m, MonadError (FinalParseError' e) m)
+  => S.Set (ErrorFancy e) -> m a
+finalFancyFailure errSet = do
+  pos <- getPosition
+  let parseErr = FancyError (pos NE.:| []) errSet
+  throwError $ ErrorFinal parseErr
+
+-- | Like 'fail', but as a "final" parse error.
+
+finalFail
+  :: (MonadParsec e s m, MonadError (FinalParseError' e) m) => String -> m a
+finalFail = finalFancyFailure . S.singleton . ErrorFail
+
+-- | Like 'customFailure', but as a "final" parse error.
+
+finalCustomFailure
+  :: (MonadParsec e s m, MonadError (FinalParseError' e) m) => e -> m a
+finalCustomFailure = finalFancyFailure . S.singleton . ErrorCustom
+
+--- * Handling errors from include files with "final" parse errors
+
+-- Some care must be taken for sources to be attached to the right parse
+-- errors when dealing with include files, so we capture the necessary
+-- procedure in this function.
+
+parseIncludeFile
+  :: forall st m a. Monad m
+  => StateT st (ParsecT CustomErr Text (ExceptT FinalParseError m)) a
+  -> st
+  -> FilePath
+  -> Text
+  -> StateT st (ParsecT CustomErr Text (ExceptT FinalParseError m)) a
+parseIncludeFile parser initState filepath text =
+  catchError parser' handler
+  where
+    parser' = do
+      eResult <- lift $ lift $
+                  runParserT (evalStateT parser initState) filepath text
+      case eResult of
+        Left parseError -> throwError $ errorFinal parseError
+        Right result -> pure result
+
+    handler e = throwError $ ErrorBundle $ attachSource filepath text e
+
+
+attachSource
+  :: FilePath -> Text -> FinalParseError' e -> FinalParseErrorBundle' e
+attachSource filePath sourceText finalParseError =
+  case finalParseError of
+    ErrorFinal parseError -> FinalParseErrorBundle'
+      { finalParseError = parseError
+      , errorSource     = sourceText
+      , sourceFileStack = filePath NE.:| []
+      }
+    ErrorBundle bundle -> bundle
+      { sourceFileStack = filePath NE.<| sourceFileStack bundle
+      }
+
+--- * Pretty-printing final parse errors
+
+-- | Pretty-print a "final" parse error: print the stack of include files,
+-- then apply the pretty-printer for custom parse errors.
+
+finalParseErrorPretty :: FinalParseErrorBundle' CustomErr -> String
+finalParseErrorPretty bundle =
+     concatMap printIncludeFile (NE.init (sourceFileStack bundle))
+  <> customParseErrorPretty (errorSource bundle) (finalParseError bundle)
+  where
+    printIncludeFile path = "in file included from " <> path <> ",\n"
+
+
```

Comment:
Phew. This stuff looks pretty tricky and opaque to a new reader. The comments are definitely appreciated.

I wonder why megaparsec stopped tracking the stack of files.

Reply:
Hmm, I think I can do better with the comments so that the code is less opaque. I will do some work on them, but it might be one or two days before I have the chance to. Is there anything  particularly opaque that I should make sure to address?

From reading the blog post and changelog for megaparsec 7, it seems that support for stacks of include files was dropped to make megaparsec faster and simpler: stacks of include files made the manipulation of source positions very hairy. Indeed, the bug we found in #853 could be an example of this, in my opinion.